xds/src/main/java/com/linecorp/armeria/xds/client/endpoint/LocalityRoutingStateFactory.java

```diff
@@ -0,0 +1,293 @@
+import com.google.common.collect.ImmutableMap;
+import com.google.protobuf.AbstractMessage;
+
+import com.linecorp.armeria.client.endpoint.EndpointGroup;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.xds.client.endpoint.XdsRandom.RandomHint;
+
+import io.envoyproxy.envoy.config.core.v3.Locality;
+
+final class LocalityRoutingStateFactory {
+
+    private static final long MOD = 10000;
+
+    private final Locality localLocality;
+
+    LocalityRoutingStateFactory(Locality localLocality) {
+        this.localLocality = localLocality;
+    }
+
+    LocalityRoutingState create(PrioritySet upstreamPrioritySet, PrioritySet localPrioritySet) {
+        // Only priority 0 is supported
+        final HostSet upstreamHostSet = upstreamPrioritySet.hostSets().get(0);
+        final HostSet localHostSet = localPrioritySet.hostSets().get(0);
+        if (upstreamHostSet == null || localHostSet == null ||
+            earlyExitNonLocalityRouting(upstreamHostSet, localHostSet, localLocality,
+                                        upstreamPrioritySet.minClusterSize())) {
+            return new LocalityRoutingState(State.NO_LOCALITY_ROUTING, localHostSet, localLocality);
+        }
+
+        final Map<Locality, EndpointGroup> upstreamHealthyPerLocality =
+                upstreamHostSet.healthyEndpointGroupPerLocality();
+        final Map<Locality, EndpointGroup> localHealthyPerLocality =
+                localHostSet.healthyEndpointGroupPerLocality();
+
+        final Map<Locality, LocalityPercentages> localityPercentages = calculateLocalityPercentages(
+                upstreamHealthyPerLocality, localHealthyPerLocality);
+
+        final LocalityPercentages localLocalityPercentages = localityPercentages.get(localLocality);
+        if (localLocalityPercentages != null && localLocalityPercentages.canRouteToUpstream()) {
+            return new LocalityRoutingState(State.LOCALITY_DIRECT, localHostSet, localLocality);
+        }
+
+        final long localPercentToRoute = localLocalityPercentages != null ?
+                                         localLocalityPercentages.localPercentToRoute() : 0;
+        final List<Locality> upstreamLocalities =
+                upstreamHealthyPerLocality.keySet().stream()
+                                          .filter(locality -> !locality.equals(localLocality))
+                                          // we assume there aren't many localities anyway, so
+                                          // we just simply sort by the string-ed value
+                                          .sorted(Comparator.comparing(AbstractMessage::toString))
+                                          .collect(Collectors.toList());
+        final ImmutableList.Builder<ResidualCapacity> residualCapacity = ImmutableList.builder();
+        // to guarantee that residualCapacity has at least one element
+        residualCapacity.add(new ResidualCapacity(localLocality, 0));
```

Comment:
So this `localLocality` will never be chosen. Is that correct?

Reply:
`localLocality` could be chosen in the extreme case if all `ResidualCapacity` have 0 

https://github.com/line/armeria/blob/885f7f818def4e80c490f91d8658b934ae884539/xds/src/main/java/com/linecorp/armeria/xds/client/endpoint/LocalityRoutingStateFactory.java#L174-L177

Having said this, the fact that this code is called implies that upstream already has healthy/degraded `localLocality` hosts so I guess this won't be a big deal.

Reply:
I missed the code. Thanks for the explanation. 👍